src/contracts/app.rs

```diff
@@ -0,0 +1,480 @@
+                error!(?e, "Failed to insert identity to contract.");
+                panic!("Failed to submit transaction, state synchronization lost.");
+                #[allow(unreachable_code)]
+                e
+            })?;
+
+        // Update  merkle tree
+        tree.set(identity_index, *commitment);
+
+        // Downgrade write lock to read lock
+        let tree = tree.downgrade();
+
+        // Check tree root
+        if let Err(error) = self.contracts.assert_valid_root(tree.root()).await {
+            error!(
+                computed_root = ?tree.root(),
+                ?error,
+                "Root mismatch between tree and contract."
+            );
+            panic!("Root mismatch between tree and contract.");
+        }
+
+        // Immediately write the tree to storage, before anyone else can write.
+        // TODO: Store tree in database
+
+        Ok(IndexResponse { identity_index })
+    }
+
+    /// # Errors
+    ///
+    /// Will return `Err` if the provided index is out of bounds.
+    #[instrument(level = "debug", skip_all)]
+    pub async fn inclusion_proof(
+        &self,
+        group_id: usize,
+        commitment: &Hash,
+    ) -> Result<InclusionProofResponse, ServerError> {
+        if U256::from(group_id) != self.contracts.group_id() {
+            return Err(ServerError::InvalidGroupId);
+        }
+        if commitment == &self.contracts.initial_leaf() {
+            return Err(ServerError::InvalidCommitment);
+        }
+
+        let merkle_tree = self.merkle_tree.read().await.map_err(|e| {
+            error!(?e, "Failed to obtain tree lock in inclusion_proof.");
+            panic!("Sequencer potentially deadlocked, terminating.");
+            #[allow(unreachable_code)]
+            e
+        })?;
+        let identity_index = match merkle_tree.leaves().iter().position(|&x| x == *commitment) {
+            Some(i) => i,
+            None => return Err(ServerError::IdentityCommitmentNotFound),
+        };
+
+        let proof = merkle_tree
+            .proof(identity_index)
+            .ok_or(ServerError::IndexOutOfBounds)?;
+        let root = merkle_tree.root();
+
+        // Locally check the proof
+        // TODO: Check the leaf index / path
+        if !merkle_tree.verify(*commitment, &proof) {
+            error!(
+                ?commitment,
+                ?identity_index,
+                ?root,
+                "Proof does not verify locally."
+            );
+            panic!("Proof does not verify locally.");
+        }
+
+        // Verify the root on chain
+        if let Err(error) = self.contracts.assert_valid_root(root).await {
+            error!(
+                computed_root = ?root,
+                ?error,
+                "Root mismatch between tree and contract."
+            );
+            return Err(ServerError::RootMismatch);
+        }
+
+        Ok(InclusionProofResponse { root, proof })
+    }
+
+    /// Stores the Merkle tree to the storage file.
```

Comment:
@recmo this comment seems incorrect?

Reply:
It is incorrect. I think this is left over from a previous function that actually did store the tree.